### PR TITLE
test: Create new test library

### PR DESCRIPTION
**Pull request by maflcko**
Sorry for clickbait, this is only a move-only scripted-diff commit and one documentation commit.

Longer term, someone who knows something about build systems can make this an actual library. Motivation for this is that each module gets compiled for each target that includes it. For example, setup_common is compiled 27 times (for the fuzz suite) and another 3 times for the other tests (bench, unit test, gui)

### CODE REVIEW

**Comment by laanwj** [comment]
I'd be interested to know what the motivation for this is, are you planning to use this library from multiple binaries, or is it just organizational?

**Comment by MarcoFalke** [comment]
> I'd be interested to know what the motivation for this is, are you planning to use this library from multiple binaries, or is it just organizational?

Yes, that is the motivation that I mentioned in the readme.

Also, we compile each file in the library (as of today) multiple times (for each target), so that should be fixed as well.

**Comment by laanwj** [comment]
> Also, we compile each file in the library (as of today) multiple times (for each target), so that should be fixed as well.

Can you give an example?

**Comment by MarcoFalke** [comment]
> Can you give an example?

setup_common is compiled 27 times (for the fuzz suite) and another 3 times for the other tests (bench, unit test, gui)

**Comment by mzumsande** [comment]
How about moving the content of `src/test/util.h` and `src/test/util.cpp` into the library as well (possibly into specific modules, much of it seems to be block and address-related)?

**Comment by laanwj** [comment]
> setup_common is compiled 27 times (for the fuzz suite) and another 3 times for the other tests (bench, unit test, gui)

Whoa. Yes, then it's quite a serious issue!

**Comment by MarcoFalke** [comment]
> How about moving the content of src/test/util.h and src/test/util.cpp into the library as well (possibly into specific modules, much of it seems to be block and address-related)?

Good point. I think I will leave this for later, so that this changeset can be kept a move-only scripted-diff

**Comment by MarcoFalke** [comment]
This pull should also fix the appveyor failures: https://github.com/bitcoin/bitcoin/pull/17357#issuecomment-550297470

File: src/test/lib/README.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# Test library
+
+This contains all files for the test library, which is used by all test binaries (unit tests, benchmarks, fuzzers, gui
+tests)
```
Comment (jonatack) [inline src/test/lib/README.md]: nit: missing "." at end of sentence

File: src/test/lib/README.md [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+Generally, the files in this folder should be well separated modules. New code should be added to existing modules or
```
Comment (jonatack) [inline src/test/lib/README.md]: s/well separated/well-separated/

File: src/test/lib/README.md [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+(when in doubt) a new module should be created.
+
+The utilities in here are compiled into a library, which does not hold any state. However, the main file `setup_common`
```
Comment (jonatack) [inline src/test/lib/README.md]: does this refer to `setup_common.cpp`, `setup_common.h`, or both?

Comment (jonatack) [inline src/test/lib/README.md]: nit: can remove the comma after "library"

**Review by jonatack — COMMENTED** [review]
Concept ACK. A few suggestions for the doc below.

**Comment by MarcoFalke** [comment]
appveyor passed: https://ci.appveyor.com/project/DrahtBot/bitcoin/builds/28656027

File: src/test/lib/README.md (hunk 3 above)
Comment (MarcoFalke) [reply src/test/lib/README.md]: both

**Comment by MarcoFalke** [comment]
Added a dot at the end of the sentence, among other minor doc fixups as requested by @jonatack 

No code changes, so the previous ci builds are still valid.

**Review by ryanofsky — APPROVED** [review]
Code review ACK fa6fb57837f7d0db90eb4bdf3b644ba540db1805. Seems good to separate test utilities from unit tests, since it's natural for things like benchmarks, fuzzing, or simulation tools to depend on test utilities, but you wouldn't want those things depending on unit tests.

Feel free to ignore this suggestion, but I think it'd be better to rename `src/test/lib` directory back to `src/test/util`. It's normal to see `lib` prefixes in library paths (especially since they get added automatically by the linker), but not really normal to see `lib/` in #include strings, and not something we have for any of our other libraries.

**Comment by MarcoFalke** [comment]
Renamed to test/util/ as requested by @ryanofsky

**Review by ryanofsky — APPROVED** [review]
Code review ACK fa4c6fa9b1139791f45f1495d662c1c7cd2f7ed6. I didn't realize `lib` was actually name of existing directory, not a new name. But in any case this looks good and nice to have one scripted diff instead of two.

**Comment by MarcoFalke** [comment]
Both travis and appveyor passes (so this fixes the appveyor build). Anything left to do here?

**Comment by practicalswift** [comment]
ACK fa4c6fa9b1139791f45f1495d662c1c7cd2f7ed6 -- diff looks correct and Travis is happy

**Comment by ryanofsky** [comment]
> Both travis and appveyor passes (so this fixes the appveyor build). Anything left to do here?

This hasn't had a ton of review, but it seems fairly safe to merge as it only affects test code.

**Comment by mzumsande** [comment]
I agree. I am concept ACK on both organizing all shared functions in one place, and on eventually turning this into a library. The code changes look good to me, but I didn't test.

Just noting that the added documentation in `src/test/util/README.md` describes the goal but not the reality after this PR, but I think fixing the Appveyor problem is more important than that.

**Comment by promag** [comment]
> Also, we compile each file in the library (as of today) multiple times (for each target), so that should be fixed as well.

And are those always compiled with the same flags etc?

**Comment by promag** [comment]
Scripted diff ACK.

File: src/test/lib/README.md (hunk 3 above)
Comment (jonatack) [reply src/test/lib/README.md]: You were right. A non-restrictive clause beginning with "that" would take no comma, but a non-restrictive clause beginning with "which" does take a comma.

**Comment by jonatack** [comment]
ACK fa4c6fa9b1139791f45f1495d662c1c7cd2f7ed6 with the reserve that the commit messages (and PR description) contain the motivation for this change. Built, ran tests, light code review.

**Comment by MarcoFalke** [comment]
Added motivation to OP as requested by @jonatack

**Pull request merged**